README.md
# Hand Gesture Controlled Robot For Human Robot Collaboration
Controlling a turtlebot with gestures so that it can be used for collaborative robots (cobots) in real time with almost zero latency between gesture recognition and publishing the commands.

[Read the Docs](https://hand-gesture-controlled-turtlebot.readthedocs.io/en/latest/about.html)


docs/changelog.rst
Changelog
=============

2.2.1
----

New:

- A performance improvement was noticed when we switched from running the ROS2 nodes in a VirtualBox to a full fledged Ubuntu 20.04 desktop in a dual boot system.


2.2.0
-------

New:

- To address the GIL limitation, we combined the three separate nodes into one Python node, which improved performance.
- However, the fastest approach was live control without crash detection functionality.
- A potential solution to overcome this limitation would be to use separate edge computers to process different types of data and inputs, with one central node making decisions.

2.1.2
-------

New:

- Previous implementation had three nodes, publishing gesture, filtered LIDAR readings and publishing Twist messages, also showing the camera view of Turtlebot 4.
- Video feed would turn red if it sensed an obstacle in proximity.
- Running multiple Python nodes caused performance issues due to Global Interpreter Lock(GIL) limitation.


2.1.1
-----

New:

- Ensured safety of the robot and avoided collisions by using readings from the Turtlebot's lidar.
- Filtered the lidar data using an FIR filter with a memory of past 50 readings due to erratic readings between subsequent readings.
- Implemented a feature to prevent the robot from moving forward if an obstacle was detected to be too close by the lidar.

.. note::

   Adjusted for the offset of the Turtlebot's lidar, which was placed at an offset of 90 degrees.
   Modified the readings to account for the offset, with the front of the lidar being considered the left side of the Turtlebot 4.
 
2.1.0
-----

New:

- Integrated the trained model into a ROS2 publisher node.
- Used the model to predict gestures from video feeds obtained from the laptop's webcam.
- Implemented a function to publish Twist messages to the /cmd/vel topic on the Turtlebot 4 based on the recognized gestures.

2.0.2
-----
New:

- Online research led to the discovery that successful solutions used relational coordinates of hand landmarks for gesture recognition instead of absolute coordinates.
- Adopting this approach yielded a model accuracy of 97%.


2.0.1
------
New:

- Trained a model using the Random Forest Classifier in scikit-learn library to recognize gestures based on landmark segment coordinates.
- Achieved a model accuracy of 85% with this approach.


2.0.0
------
New:

- The previous approach was ineffective due to the model's size and lack of improvement in accuracy and latency.
- Mediapipe python library was adopted as an alternative solution, leveraging 21 hand landmarks to segment human hands. More information on mediapipe python library can be found here: https://developers.google.com/mediapipe/solutions/vision/hand_landmarker.
- The X and Y coordinates of the landmarks can be recorded and used to train a machine learning model for gesture recognition.

1.0.2
-------
New:

- The model was trained on isolated hand images obtained from a dataset on Kaggle found in https://www.kaggle.com/datasets/muhammadkhalid/sign-language-for-numbers?resource=download.
- Accurate predictions required segmentation of hands from the video feed.
- Testing the model on live video feed revealed a significant latency between video input and gesture recognition.
- The latency was unsuitable for real-time control of the Turtlebot4, hindering its use as a remote-controlled collaborative robot.

1.0.1
-------
New:

- Technical challenges arose while training the model on Virtualbox running Ubuntu 20.04 due to the absence of AVX and AVX2 instruction sets required by Tensorflow.
- A possible workaround is to use Tensorflow Lite instead of Tensorflow.
- The large size of the model prompted a need to reduce inference time, leading to the decision to dualboot laptops with Ubuntu 20.04 and continue using Tensorflow.


1.0.0
-------
New:

- We plan on training a Convolutional Neural Network (CNN) using Tensorflow library for the gesture detection mechanism.
- We plan on creating the CNN model using Tensorflow and feeding it images found on Kaggle dataset.
- The Model will be trained on a vast dataset containing images of people and hands in various positions, backgrounds and also have some unknown set of images to train the network.

To achieve the above we follow the steps given below:

Created a new environment named 'tf' with Python 3.9 using the following command:

.. code-block:: console

   $ conda create --name tf python=3.9

Deactivated the current environment and activated the newly created environment using the following commands:

.. code-block:: console

   $ conda deactivate

   $ conda activate tf

Installed the necessary libraries for GPU support using the following commands:

.. code-block:: console

   $ conda install -c conda-forge cudatoolkit=11.2 cudnn=8.1.0

Upgraded pip to the latest version using the following command:

.. code-block:: console


   $ pip install --upgrade pip

Installed tensorflow version 2.10 using the following command:

.. code-block:: console

   $ pip install "tensorflow<2.11"

Note: Anything above 2.10 is not supported on the GPU on Windows Native.

Verified the installation by running the following commands:

.. code-block:: console

   $ python -c "import tensorflow as tf; print(tf.reduce_sum(tf.random.normal([1000, 1000])))"
   $ python -c "import tensorflow as tf; print(tf.config.list_physical_devices('GPU'))"

The first command checks if tensorflow can run a simple operation on the GPU, while the second command checks if the GPU is detected by tensorflow.
If tensorflow was installed correctly then it should list the GPU on the device.

More details on installing tensorflow on Windows can be found here: https://www.tensorflow.org/install/pip#windows-native_1 


RF2_media_video_test.py
import numpy as np
import cv2
import mediapipe as mp
import pickle

# Define the list of landmark names to be used
# landmark_names = ['cx1', 'cx2', 'cx3', 'cx4', 'cx5', 'cx6', 'cx7', 'cx8', 'cx9', 'cx10', 'cx11', 'cx12', 'cx13', 'cx14', 'cx15', 'cx16', 'cx17', 'cx18', 'cx19', 'cx20', 'cx21', 'cy1', 'cy2', 'cy3', 'cy4', 'cy5', 'cy6', 'cy7', 'cy8', 'cy9', 'cy10', 'cy11', 'cy12', 'cy13', 'cy14', 'cy15', 'cy16', 'cy17', 'cy18', 'cy19', 'cy20', 'cy21']
# model_dict = pickle.load(open(fr"C:\Users\Archit Jain\OneDrive - Arizona State University\Desktop\ASU Classes\EGR 598 Exp and Depl of Robotics\Project\model_1.p", 'rb'))
model_dict = pickle.load(open(fr"C:\Users\Archit Jain\OneDrive - Arizona State University\Desktop\ASU Classes\EGR 598 Exp and Depl of Robotics\Project\Sign_Language_Grey_classifier_2.p", 'rb'))



model = model_dict['model']
cap = cv2.VideoCapture(0)
key = [0,1,2,3,4,5,6,7,8,9]


mp_hands = mp.solutions.hands
mp_drawing = mp.solutions.drawing_utils
mp_drawing_styles = mp.solutions.drawing_styles

hands = mp_hands.Hands(static_image_mode=True, min_detection_confidence=0.7)

while True:
    data_aux = []
    x_ = []
    y_ = []

    ret, image = cap.read()
    if not ret:
            break
    H, W, _ = image.shape
    
    # Detection implementation #############################################################
    image = cv2.cvtColor(image, cv2.COLOR_BGR2RGB)

    # # Process the current frame with Mediapipe Hands model
    results = hands.process(image)

    if results.multi_hand_landmarks:
        for hand_landmarks in results.multi_hand_landmarks:
            # Extract the cx and cy coordinates of the landmarks and store them in a list
            mp_drawing.draw_landmarks(
                image,  # image to draw
                hand_landmarks,  # model output
                mp_hands.HAND_CONNECTIONS,  # hand connections
                mp_drawing_styles.get_default_hand_landmarks_style(),
                mp_drawing_styles.get_default_hand_connections_style())

        for hand_landmarks in results.multi_hand_landmarks:
            for i in range(len(hand_landmarks.landmark)):
                x = hand_landmarks.landmark[i].x
                y = hand_landmarks.landmark[i].y

                x_.append(x)
                y_.append(y)

            for i in range(len(hand_landmarks.landmark)):
                x = hand_landmarks.landmark[i].x
                y = hand_landmarks.landmark[i].y
                data_aux.append(x - min(x_))
                data_aux.append(y - min(y_))

        x1 = int(min(x_) * W) - 10
        y1 = int(min(y_) * H) - 10

        x2 = int(max(x_) * W) - 10
        y2 = int(max(y_) * H) - 10

        prediction = model.predict([np.asarray(data_aux)])

        predicted_character = key[int(prediction[0])]

        cv2.rectangle(image, (x1, y1), (x2, y2), (0, 0, 0), 4)
        cv2.putText(image, str(predicted_character), (x1, y1 - 10), cv2.FONT_HERSHEY_SIMPLEX, 1.3, (0, 0, 0), 3,
                     cv2.LINE_AA)

    cv2.imshow('frame', image)
    cv2.waitKey(1)


cap.release()
cv2.destroyAllWindows()








cobot3_ws/src/cobot/cobot/drive.py
import rclpy
from rclpy.node import Node
import cv2
import numpy as np
from geometry_msgs.msg import Twist
import mediapipe as mp
import time
from std_msgs.msg import Int32, Float32
from sensor_msgs.msg import Image, LaserScan
from cv_bridge import CvBridge
import matplotlib.pyplot as plt 
from matplotlib.animation import FuncAnimation
import numpy
from scipy.signal import convolve, firwin


from geometry_msgs.msg import Twist
import pickle
import mediapipe as mp
import time
MEMORY = 50

my_frame  = numpy.zeros((MEMORY,2))


class TBDriver(Node):
    def __init__(self):
        super().__init__('subscriber')
        # self.subscription = self.create_subscription(Image,'/color/preview/image', self.listener_callback, 1)
        self.publisher = self.create_publisher(Twist, '/cmd_vel', 10)
        self.twist_msg = Twist()
        self.twist_msg.linear.x = 0.0  # 0.0 m/s
        self.twist_msg.angular.z = 0.4  # 0.4 rad/s
        
        # self.gest_subscription = self.create_subscription(
        #         Int32,
        #         '/gesture',
        #         self.gesture_callback,
        #         1)
        
        self.scan_subscribtion = self.create_subscription(
            LaserScan,
            '/scan',
            self.scan_callback,
            1)
        self.scan_subscribtion
        self.t0 = None
        self.history = None
        self.fig  = None
        self.i = 0
        
        self.image_subscription = self.create_subscription(
            Image,
            '/color/preview/image',
            self.image_callback,
            1)
        
        self.bridge = CvBridge()
        
        self.model_dict = pickle.load(open("/home/architjain/Downloads/Sign_Language_Grey_classifier_2.p", 'rb'))

        self.model = self.model_dict['model']
        self.cap = cv2.VideoCapture(0)
        self.key = [0,1,2,3,4,5,6,7,8,9]


        self.mp_hands = mp.solutions.hands
        self.mp_drawing = mp.solutions.drawing_utils
        self.mp_drawing_styles = mp.solutions.drawing_styles

        self.hands = self.mp_hands.Hands(static_image_mode=True, min_detection_confidence=0.7)
        timer_period = 0.1  # seconds
       
        # Create the timer
        self.timer = self.create_timer(timer_period, self.timer_callback)

        # initialize random variables
        self.front_range = 10.097986976
        self.gesture = 9
        self.tb_view = np.zeros((480, 640, 3), dtype=np.uint8)

        
    def timer_callback(self):
        data_aux = []
        x_ = []
        y_ = []

        ret, image = self.cap.read()
        
        H, W, _ = image.shape
        
        # Detection implementation #############################################################
        image = cv2.cvtColor(image, cv2.COLOR_BGR2RGB)

        # image.flags.writeable = False
        # # results = hands.process(image)
        # image.flags.writeable = True


        # # Initialize the Mediapipe Hands model
        # mp_hands = mp.solutions.hands
        # hands = mp_hands.Hands(static_image_mode=False, max_num_hands=2, min_detection_confidence=0.85)

        # # Process the current frame with Mediapipe Hands model
        results = self.hands.process(image)

        if results.multi_hand_landmarks:
            count = 0
            for hand_landmarks in results.multi_hand_landmarks:
                # Extract the cx and cy coordinates of the landmarks and store them in a list
                if count < 1:
                    self.mp_drawing.draw_landmarks(
                        image,  # image to draw
                        hand_landmarks,  # model output
                        self.mp_hands.HAND_CONNECTIONS,  # hand connections
                        self.mp_drawing_styles.get_default_hand_landmarks_style(),
                        self.mp_drawing_styles.get_default_hand_connections_style())
                    count += 1
            count = 0
            for hand_landmarks in results.multi_hand_landmarks:
                if count < 1:
                    for i in range(len(hand_landmarks.landmark)):
                        x = hand_landmarks.landmark[i].x
                        y = hand_landmarks.landmark[i].y

                        x_.append(x)
                        y_.append(y)

                    for i in range(len(hand_landmarks.landmark)):
                        x = hand_landmarks.landmark[i].x
                        y = hand_landmarks.landmark[i].y
                        data_aux.append(x - min(x_))
                        data_aux.append(y - min(y_))
                    count += 1

            x1 = int(min(x_) * W) - 10
            y1 = int(min(y_) * H) - 10

            x2 = int(max(x_) * W) - 10
            y2 = int(max(y_) * H) - 10

            prediction = self.model.predict([np.asarray(data_aux)])

            predicted_character = self.key[int(prediction[0])]

            cv2.rectangle(image, (x1, y1), (x2, y2), (0, 0, 0), 4)
            cv2.putText(image, str(predicted_character), (x1, y1 - 10), cv2.FONT_HERSHEY_SIMPLEX, 1.3, (0, 0, 0), 3,
                        cv2.LINE_AA)

            print(predicted_character)
            self.gesture = int(predicted_character)
            
            

        cv2.imshow('User Video', image)
        cv2.waitKey(1)


    def publish_twist(self):
        self.publisher.publish(self.twist_msg)
        self.get_logger().info('Publishing Twist message: linear=%f, angular=%f' %
                                    (self.twist_msg.linear.x, self.twist_msg.angular.z))
        
    def scan_callback(self, msg):
            # Get the range at the center of the scan
        range_index = len(msg.ranges) // 4
        range_value = msg.ranges[range_index]

        # Print the range in front of the robot
        self.get_logger().info(f'Range in front of the robot: {range_value:.6f}')
        current_time = msg.header.stamp.sec+msg.header.stamp.nanosec*1e-9
        if not self.t0:
            self.t0 = current_time

        t = current_time-self.t0


        if self.history is None:
            self.history = numpy.zeros((MEMORY,2))

        self.history[:-1,:] = self.history[1:,:]
        self.history[-1,:] = numpy.array([t, range_value])
        self.update(self.i)

    def update(self,i):

        if self.fig is None:
            first = True
            self.fig = plt.figure()
            self.ax = plt.subplot(111)
        else:
            first = False
        

        # #Apply FIR Filter
        x = self.history[:,1]
        time = self.history[:, 0]

        # Define the filter coefficients using the firwin function
        filter_order = 10
        cutoff_freq = 0.1
        b = firwin(filter_order, cutoff_freq)

        # Apply the filter using the convolve function
        y = convolve(x, b, mode='valid')

        
        self.front_range = y[-1]
        self.get_logger().info(f'Range in front: {y[-1]:.2f}')

        # Plot Everything
        self.ax.cla()
        self.ax.plot(self.history[filter_order:,0],self.history[filter_order:,1], label="Scan Range")
        self.ax.plot(time[filter_order-1:], y, label = "FIR Filtered Data")
        # self.ax.set_ylim(self.history[:,1].min(),self.history[:,1].max())    # plot memory
        self.fig.canvas.draw()
        if first:
            plt.ion()
            plt.show()
    
    def gesture_callback(self, msg):
            self.gesture = msg.data
    
    def image_callback(self, msg):
        # Store the image data from the /image topic
        self.tb_view = self.bridge.imgmsg_to_cv2(msg)

    def process_data(self):
        # Do something with the data from all three topics
        
    
        tb_view2 = self.tb_view
        
        if self.gesture == 1:
            if self.front_range > 0.35:
                self.twist_msg.linear.x = 0.2
                self.twist_msg.angular.z = 0.0
            else:
                self.twist_msg.linear.x = 0.0
                self.twist_msg.angular.z = 0.0
                tb_view2[:,:,2] = np.clip(tb_view2[:,:,2] + 100, 0, 255)
                # Add text to the image
                font = cv2.FONT_HERSHEY_SIMPLEX
                bottomLeftCornerOfText = (10, 50)
                fontScale = 1
                fontColor = (0, 0, 255) # red
                lineType = 2
                cv2.putText(tb_view2, 'Robot will crash into object', 
                    bottomLeftCornerOfText, font, fontScale, fontColor, lineType)

        elif self.gesture == 2:
            self.twist_msg.linear.x = -0.2
            self.twist_msg.angular.z = 0.0
        elif self.gesture == 3:
            self.twist_msg.linear.x = 0.0
            self.twist_msg.angular.z = -0.4
        elif self.gesture == 4:
            self.twist_msg.linear.x = 0.0
            self.twist_msg.angular.z = 0.4
        elif self.gesture == 5:
            self.twist_msg.linear.x = 0.0
            self.twist_msg.angular.z = 0.0
        self.publish_twist()
        cv2.imshow('Turtlebot Live View', tb_view2)
        cv2.waitKey(1)



def main(args=None):
    rclpy.init(args=args)
    tbdriver = TBDriver()
    
    plt.ion()
    plt.show()
    
    while rclpy.ok():
        tbdriver.process_data()
        rclpy.spin_once(tbdriver)

    tbdriver.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()